crypto/hpke/hpke_util.c

```diff
@@ -173,3 +396,72 @@ EVP_KDF_CTX *ossl_kdf_ctx_create(const char *kdfname, const char *mdname,
     }
     return kctx;
 }
+
+/*
+ * @brief map a string to a HPKE suite based on synonym tables
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ */
+int ossl_hpke_str2suite(const char *suitestr, OSSL_HPKE_SUITE *suite)
```

Comment:
This whole methods is a bit over complicated.
Simplified and improved version : 
````C

/*
 * @brief look for a label into the synonym tables, and return its id
 * @param st is the string value
 * @param synp is the synonyms labels array 
 * @param outsize is the previous array size
 * @return 0 when not found, else the matching item id.
 */
static uint16_t synonyms_lookup(const char st, const synonymttab_t *synp, size_t outsize)
{
    size_t i, j;

    for (i = 0; i < outsize; ++i) {
        for (j = 0; j < OSSL_NELEM(synp[i].synonyms); ++j) {
            if (OPENSSL_strcasecmp(st, synp[i].synonyms[j]) == 0)
                return synp[i].id;
        }
    }
    return 0;
}

/*
 * @brief map a string to a HPKE suite based on synonym tables
 * @param str is the string value
 * @param suite is the resulting suite
 * @return 1 for success, otherwise failure
 */
int ossl_hpke_str2suite(const char *suitestr, OSSL_HPKE_SUITE *suite)
{
    uint16_t kem = 0, kdf = 0, aead = 0, res;
    char *st, char *instrcp;
    size_t inplen;
    int labels = 0, result = 0;

    if (suitestr == NULL || suitestr[0] == 0x00 || suite == NULL) {
        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
        return 0;
    }

    inplen = OPENSSL_strnlen(suitestr, OSSL_HPKE_MAX_SUITESTR);
    if (inplen >= OSSL_HPKE_MAX_SUITESTR)
        return 0;
    /* Duplicate `suitestr` to allow its parsing  */
    instrcp = OPENSSL_memdup(suitestr, inplen + 1);
    if (instrcp == NULL)
        goto fail;

    /* See if it contains a mix of our strings and numbers */
    st = strtok(instrcp, ",");
    if (st == NULL)
        goto fail;

    while (st != NULL && labels < 3) {
        /* check if string is known or number and if so handle appropriately */
        if (labels == 0
            && (kem = synonyms_lookup(kemstrtab, OSSL_NELEM(kemstrtab))) == 0)
                goto fail;
        else
        if (labels == 1
            && (kdf = synonyms_lookup(kdfstrtab, OSSL_NELEM(kdfstrtab))) == 0)
            goto fail;
        else
        if (labels == 2
            && (aead = synonyms_lookup(aeadstrtab, OSSL_NELEM(aeadstrtab))) == 0)
            goto fail;

        st = strtok(NULL, ",");
        ++labels;
    }
    suite->kem_id = kem;
    suite->kdf_id = kdf;
    suite->aead_id = aead;
    result = 1;

 fail:
    OPENSSL_free(instrcp);
    return result;
}
````

Reply:
I kinda like that better yeah, (it's clearer so should be easier to maintain). Will test it and get back, but plan to change to that 'less someone tells me not to do that.

Reply:
with a couple of tweaks that seems to work fine and not make valgrind unhappy - I'll try it in my ECH fork and see if any of the CI builds get upset by it before committing (will be >1 hour as I'll also be in transit meantime)

Reply:
I don't compile this code, it have one mistake at least  : 
`synonyms_lookup` calls are missing the `st` string as 1st argument

Reply:
no worries, was easily fixed to compile & work & pass tests

Reply:
so I adopted this (with a few changes as noted); I also added a few more test cases for that as well.